source/simulator/core.cc

```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+                     || introspection.variable("compaction pressure").block_index == 1, ExcNotImplemented());
+
+              // create a completely distributed vector that will be used for
+              // the scaled and denormalized solution and later used as a
+              // starting guess for the linear solver
+              LinearAlgebra::BlockVector linearized_stokes_initial_guess (introspection.index_sets.stokes_partitioning, mpi_communicator);
+
+              linearized_stokes_initial_guess.block (block_vel) = current_linearization_point.block (block_vel);
+              linearized_stokes_initial_guess.block (block_p) = current_linearization_point.block (block_p);
+
+              if (nonlinear_iteration == 0)
+                {
+                  initial_residual = this->compute_initial_newton_residual(linearized_stokes_initial_guess);
+                  switch_initial_residual = initial_residual;
+                  residual = initial_residual;
+                }
+
+              assemble_newton_stokes_system = assemble_newton_stokes_matrix = true;
+
+              denormalize_pressure (this->last_pressure_normalization_adjustment,
+                                    linearized_stokes_initial_guess,
+                                    current_linearization_point);
+
+              compute_current_constraints ();
+
+              // the Stokes matrix depends on the viscosity. if the viscosity
+              // depends on other solution variables, then after we need to
+              // update the Stokes matrix in every time step and so need to set
+              // the following flag. if we change the Stokes matrix we also
+              // need to update the Stokes preconditioner.
+              rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+
+              assemble_stokes_system();
+
+              /**
+               * Eisenstat Walker method for determining the tolerance
+               */
+              if (nonlinear_iteration > 1)
+                {
+                  residual_old = residual;
+                  velocity_residual = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+                  pressure_residual = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  residual = std::sqrt(velocity_residual * velocity_residual + pressure_residual * pressure_residual);
+
+                  if (!use_picard)
+                    {
+                      const bool EisenstatWalkerChoiceOne = true;
+                      this->compute_Eisenstat_Walker_linear_tolerance(EisenstatWalkerChoiceOne,
+                                                                      parameters.maximum_linear_stokes_solver_tolerance,
+                                                                      parameters.linear_stokes_solver_tolerance,
+                                                                      stokes_residual,
```

Comment:
you are right, I moved the initialisation a outside the loop and it has (in the few test I did) significantly improved the convergence! :)